server/src/main/java/io/crate/geo/GeoJSONUtils.java

```diff
@@ -155,6 +181,81 @@ public static Object map2LuceneShape(Map<String, Object> geoJSONMap) {
         return geoJSONString2ShapeBuilder(geoJSONMap).buildLucene();
     }
 
+    /**
+     * Checks equality, returns 0 for equal shapes and 1 for non-equal.
+     * This method doesn't do any validation and assumes that Maps represent correct geo shapes.
+     * When comparing geometries collections, order of geometries is important,
+     * ie collection(point1, point2) != collection(point2, point1).
+     */
+    @SuppressWarnings("unchecked")
+    public static int compare(Map<String, Object> val1, Map<String, Object> val2) {
+        Object type1 = val1.get(TYPE_FIELD);
+        Object type2 = val1.get(TYPE_FIELD);
+        assert type1 != null && type2 != null : "Shapes must be valid"; // implicitCast and sanitizeValue validate provided values.
+        if (type1.equals(type2) == false) {
+            return 1;
+        }
+
+        int result;
+        if (GEOMETRY_COLLECTION.equals(type1)) {
+            // Both shapes are valid geometries collections.
+            Object geometries1 = val1.get(GeoJSONUtils.GEOMETRIES_FIELD);
+            Object geometries2 = val2.get(GeoJSONUtils.GEOMETRIES_FIELD);
+            Iterator<?> it1 = iterator(geometries1);
+            Iterator<?> it2 = iterator(geometries2);
+            result = 0;
+            while (it1.hasNext()) {
+                if (it2.hasNext() == false) {
+                    // Geometry collections of different sizes are not equal (collection1.size > collections2.size).
+                    result = 1;
+                    break;
+                }
+                Object value1 = it1.next();
+                Object value2 = it2.next();
+                assert value1 instanceof Map : "Shapes must be valid";
+                assert value2 instanceof Map : "Shapes must be valid";
+                result += compare((Map<String, Object>) value1, (Map<String, Object>) value2);
+                if (result > 0) {
+                    break;
+                }
+            }
+            if (it2.hasNext()) {
+                // Geometry collections of different sizes are not equal (collection2.size > collections1.size).
+                result = 1;
+            }
+
+        } else {
+            Object coordinates1 = val1.get(COORDINATES_FIELD);
+            Object coordinates2 = val2.get(COORDINATES_FIELD);
+            if (coordinates1 instanceof double[] arr1 && coordinates2 instanceof double[] arr2) {
+                // Not using compare as it can return -1, geo shape compare returns either 0 or 1.
+                return Arrays.equals(arr1, arr2) ? 0 : 1;
+            }
```

Comment:
What about linestring, polygon, multipolygon which are `double[][]`, `double[][][]` and `double[][][][]` ?

Reply:
I will add those cases, thanks. Currently investigating why we end up with Lists here. 

`ShardUpsertRequest.Item` sometimes has primitives but at some point it has Lists, figuring out where this conversion happens.

Reply:
Ok, it's GeoShapeType.readValueFrom -> readGenericValue -> readArray. Turns double[] into List.

This is not related to this PR, ie I will add extra cases as we can have them when not streaming and array-> List should be handled separately.